app/gui2/shared/yjs-model.ts

```diff
@@ -232,6 +232,13 @@ class DistributedModule {
   dispose(): void {
     this.doc.destroy()
   }
+
+  updateCode(code: string): void {
+    this.doc.transact(() => {
+      this.contents.delete(0, this.contents.length)
+      this.contents.insert(0, code)
```

Comment:
You should not do text updates like this. It will not play nicely with networked updates. Each update will effectively destroy any changes other peers have made in the mean time. It will also completely trash all assigned AST IDs, as those are stored as [Yjs relative positions](https://docs.yjs.dev/api/relative-positions). A large text replace like this one will move all existing relative positions to the beginning of the text.

The solution here is to actually use the `beforeupdate` event, and manually apply text edits to ranges reported by it. Alternatively, you can actually go all the way and use an existing code editor like monaco, which conveniently already has an integration with YJs.

https://docs.yjs.dev/ecosystem/editor-bindings/monaco